### ISSUE TITLE
Failures on Generic metadata test.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#45120]
There are new test failures on [Generic metadata test.](https://dart-review.googlesource.com/c/sdk/+/185583).

The tests
```
language/generic/generic_metadata_test CompileTimeError (expected Pass)
```
are failing on configurations
```
cfe-strong-linux
cfe-weak-linux
```

**Comment by johnniwinther** [comment]
cc @jensjoha 

Several problems in the file.

1) We don't handle type arguments on annotations in typedefs. In line 211 we have problems with these two type arguments:
```
@A<int>(0)
typedef F<@A<int>(0) T> = int Function<@A<int>(0) X>(@A<int>(0) int);
            ^^^^^                        ^^^^^
```

2) We don't handle type arguments on type parameters in function type annotation (maybe that's the problem with the second annotation above).  In line 228 we have a problem with these type arguments:
```
  @A<int>(0)
  void Function<@A<int>(0) X>(@A<int>(0) int y)? fv = null;
                  ^^^^^
```

3) When removing the problematic type arguments in line 228 we emit the following error:
```
tests/language/generic/generic_metadata_test.dart:228:23: Error: A type variable on a function type can't have annotations.
  void Function<@A(0) X>(@A<int>(0) int y)? fv = null;
                      ^
```

@eernstg Should we allow annotations on type parameters of function type?

**Comment by eernstg** [comment]
> Should we allow annotations on type parameters of function type?

Yes, cf. https://github.com/dart-lang/sdk/issues/43039, and the generic ones like `void Function<@A<int>(0) X>(T)` would be supported as well, just because there is no special restriction on that kind of annotations.

However, we do not promise that constant evaluation will take place for an annotation on a type parameter or value parameter in a function type. They need to be type correct, but if the evaluation would throw then we might not get an error for that.
